### ISSUE TITLE
Lightweight rich text editing for comments (and maybe more)

### ISSUE THREAD

**Issue by sz332** [gaphor/gaphor#2582]
### Is your feature request related to a problem?

It would be nice to have rich text editing capability. This could be used for documentation, comments, and every other place where it makes sense (requirement description for example).

### Describe the solution you'd like

When modifying such a text, I would like to see a toolbar where I can set the following to the text part I select:

- Bold
- Italic
- Bullet list
- Numbered list
- Link to external URL: rendered as hyperlink, when clicked, opened in the OS default browser)
- Link to existing model element: name is displayed, rendered as link, when clicked on the link we just there in the tree, when name is modified in the model, it will be also visible in there

**Comment by sz332** [comment]
It seems that GTK's TextView actually supports this capability on some level:

https://askubuntu.com/questions/272446/pygtk-textbuffer-adding-tags-and-reading-text

**Comment by mikekidner** [comment]
The styling of the text elements should reference the CSS Stylesheet, so perhaps an HTML or markdown syntax should underpin this. Perhaps similar to this comment box.

**Comment by sz332** [comment]
@mikekidner Yeah, I was thinking about that same. How it is being implemented I would keep it open at this point of the discussion.

**Comment by amolenaar** [comment]
Rendering links can be tricky, since the diagram is basically an image.

I suggest markdown format. It is easy to save and easy to render.

**Comment by mikekidner** [comment]
Supporting markdown would be great

A possible behaviour could be:
Text items display as rendered markdown, when the uses clicks to edit the item in the property editor it displays in raw markdown.
For now this would rely on the user being familiar with markdown.
The addition of local format controls [H, B, I <> etc] could be added later.
<img width="650" alt="Screen Shot 2023-08-10 at 7 19 01 am" src="https://github.com/gaphor/gaphor/assets/50702106/8e59dfcd-f517-4f32-9532-e6bc698b0004">

The element.note field and the comment block would be good places to start.
 
I'm currently writing some tools to pull data from models such as system descriptions (hopefully these will form a gaphor plugin at some point...) currently I export to markdown, so the markdown in comments and notes will come through and be rendered after export.

**Comment by amolenaar** [comment]
> For now this would rely on the user being familiar with markdown.

We could add a little help button, like we have for class attributes and operations, explaining the syntax a bit.

**Comment by sz332** [comment]
Well, we could even add action buttons, similar to what we have it in github, also later we could add something like Preview, the same way as github does.

**Comment by ghillebrand** [comment]
+1 from me - I would love to see this feature as one builds models aimed at less technical users, and capturing richer information. ie UML models aimed outward of the programming/ tech community, rather than inward.

**Comment by amolenaar** [comment]
Pango, the library we use for text rendering, already has [some support for markup](https://docs.gtk.org/Pango/pango_markup.html). If you know of a good markdown parser 😉, and let it render to a Pango layout object.

**Comment by ghillebrand** [comment]
I'm realising that the issue is broader than just _having_ markdown, but how (if/ when) which text fields are displayed in models, which may be a standards thing. At analysis/ design time, one has different needs to technical details. The recently completed SysML [references ](https://github.com/gaphor/gaphor/issues/2967) feature touched on this - being able to show the notes text. I'm not sure where that discussion sits on the roadmap!

As a library, [Python-Markdown](https://github.com/Python-Markdown/markdown) ?
I don't understand the details of a BSD 3-Clause License vs Apache2 - is there a conflict?
